### PR TITLE
cryptsetup: measure LUKS volume keys to PCR 15

### PR DESCRIPTION
**Pull request by poettering**
See: #24503

### CODE REVIEW

**Comment by bluca** [comment]
Is this material for the next release? Or does it need to be paired with the other PCR changes in 252?

**Comment by poettering** [comment]
> Is this material for the next release? Or does it need to be paired with the other PCR changes in 252?

Next release I guess.

I mean, this is fairly isolated, but it's also new stuff, and we are in rc freeze, so let's follow our rules and not merge new stuff anymore for this cycle

File: man/systemd-pcrphase.service.xml [hunk 1]
```diff
@@ -131,6 +156,21 @@
         all suitable TPM2 devices currently discovered.</para></listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--machine-id</option></term>
+
+        <listitem><para>Instead of measuing a word specified on the command line into PCR 11, measure the
```
Comment (behrmann) [inline man/systemd-pcrphase.service.xml]:
```suggestion
        <listitem><para>Instead of measuring a word specified on the command line into PCR 11, measure the
```

File: man/systemd-pcrphase.service.xml [hunk 2]
```diff
@@ -131,6 +156,21 @@
...
+        host's machine ID into PCR 15.</para></listitem>
+      </varlistentry>
+
+      <varlistentry>
+        <term><option>--file-system=</option></term>
+
+        <listitem><para>Instead of measuing a word specified on the command line into PCR 11, measure
```
Comment (behrmann) [inline man/systemd-pcrphase.service.xml]:
```suggestion
        <listitem><para>Instead of measuring a word specified on the command line into PCR 11, measure
```

File: man/systemd.mount.xml [hunk 3]
```diff
@@ -366,6 +366,18 @@
...
+      <varlistentry>
+        <term><option>x-systemd.pcrfs</option></term>
+
+        <listitem><para>Measures file system identity information (mount point, type, label, UUID, partition
+        label, partition UUID) into PCR 15 after the file system has been mounted. This ensures the
+        <citerefentry><refentrytitle>systemd-pcrfs@.service</refentrytitle><manvolnum>8</manvolnum></citerefentry>
+        or <filename>systemd-pcrfs-root.service</filename> services are pulled in by the mount unit.</para>
+
+        <para>Note that this option can only be used in <filename>/etc/fstab</filename>, and will be ignored
+        when part of the <varname>Options=</varname> setting in a unit file.</para></listitem>
```
Comment (behrmann) [inline man/systemd.mount.xml]: How does this mesh with the discoverable partition spec, e.g. on a system without fstab?

Comment (poettering) [reply man/systemd.mount.xml]: gpt-auto-generator also hooks this up automatically, see the gpt-auto-generator patch in this PR

Comment (poettering) [reply man/systemd.mount.xml]: (will add comment to man page saying this explicitly)

Comment (bluca) [reply man/systemd.mount.xml:378]: @poettering MEETING

**Comment by keszybz** [comment]
CI is unhappy:
```
[   29.945717] systemctl[420]: Failed to switch root: Specified switch root path '/sysroot' does not seem to be an OS tree. os-release file is missing.
[   29.948249] systemctl[420]: Bus n/a: changing state RUNNING → CLOSED
Bus private-bus-connection: changing state RUNNING → CLOSING
Bus private-bus-connection: changing state CLOSING → CLOSED
Got disconnect on private connection.
Received SIGCHLD from PID 420 (systemctl).
Child 420 (systemctl) died (code=exited, status=1/FAILURE)
initrd-switch-root.service: Child 420 belongs to initrd-switch-root.service.
initrd-switch-root.service: Main process exited, code=exited, status=1/FAILURE
initrd-switch-root.service: Failed with result 'exit-code'.
initrd-switch-root.service: Service will not restart (restart setting)
initrd-switch-root.service: Changed start -> failed
initrd-switch-root.service: bpf-lsm: Failed to delete cgroup entry from LSM BPF map: No such file or directory
initrd-switch-root.service: Job 74 initrd-switch-root.service/start finished, result=failed
[FAILED] Failed to start initrd-switch-root.service.
See 'systemctl status initrd-switch-root.service' for details.
```

**Comment by poettering** [comment]
Force pushed a new version that hopefully addresses this.

**Comment by poettering** [comment]
(dont't bother to review for now, fighting with some CI test issues)

**Comment by poettering** [comment]
Hmm, apparently all CI issues are gone now, so unsetting dont-merge again. I must have forgotten a push after some earlier fixing...

File: src/shared/tpm2-util.c [hunk 4]
```diff
@@ -1957,8 +1965,15 @@ int tpm2_extend_bytes(
 
                 values.digests[values.count].hashAlg = id;
 
-                if (EVP_Digest(data, sz, (unsigned char*) &values.digests[values.count].digest, NULL, implementation, NULL) != 1)
-                        return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE), "Failed to hash word.");
+                /* So here's a twist: sometimes we want to measure secrets (e.g. root file system volume
+                 * key), but we'd rather not leak a literal hash of the secret to the TPM (given that the
+                 * wire is unprotected). Hence we instead measure an HMAC signature of a non-secret string
+                 * instead. */
```
Comment (keszybz) [inline src/shared/tpm2-util.c]: I was trying to think if there are cases where this "leak" would make a difference — it cannot be used to recover the original data, and something like a volume key would also be different between systems. I guess it'd be nice if the justification of what we're trying to protect against was made a bit stronger.

File: src/cryptsetup/cryptsetup.c [hunk 5]
```diff
@@ -762,6 +801,149 @@ static int get_password(
...
+
+        _cleanup_(erase_and_freep) void *vk = NULL;
+        size_t vks;
+        int r;
+
+        assert(cd);
+
+        /* A wrapper around crypt_activate_by_passphrase() which also measures to a PCR if that's
+         * requested. Note that we need the volume key for the measurement, and
+         * crypt_activate_by_passphrase() doesn't give us access to us. Hence, we operate indirectly, and
```
Comment (keszybz) [inline src/cryptsetup/cryptsetup.c]: "to us" → "to this"?

File: man/crypttab.xml [hunk 6]
```diff
@@ -700,6 +700,28 @@
...
+      <varlistentry>
+        <term><option>tpm2-measure-pcr=</option></term>
+
+        <listitem><para>Controls whether to measure the volume key of the encrypted volume to a TPM2 PCR. If
+        set to "no" (which is the default) no PCR extension is done. If set to "yes" the volume key is
+        measured into PCR 15. If set to a decimal integer in the range 0…23 the volume key is measured into
+        the specified PCR. The volume key is measured along with the activated volume name and its UUID. This
+        functionality is particularly useful for the encrypted volume backing the root file system, as it
+        then allows later TPM objects to be securely bound to root file system and hence installed
+        system.</para></listitem>
```
Comment (keszybz) [inline man/crypttab.xml]: "to the root file system and hence the specific installation" ?

File: man/crypttab.xml [hunk 7]
```diff
@@ -700,6 +700,28 @@
...
+      </varlistentry>
+
+      <varlistentry>
+        <term><option>tpm2-measure-bank=</option></term>
+
+        <listitem><para>Selects the TPM2 PCR bank to measure the volume key into, as configured with
+        <option>tpm2-measure-pcr=</option> above. May be used multiple times to configure multiple PCR banks
+        to measure into. If not specified automatically determines available and used banks. Expects a
+        message digest name (e.g. <literal>sha1</literal>, <literal>sha256</literal>, …) as argument, to
+        identify the bank.</para></listitem>
```
Comment (keszybz) [inline man/crypttab.xml]: If we expect people to measure into more than one bank, it'd be nice to allow multiple banks to be specified with one argument. It's going to be painfully verbose to specify more than two with a separate (lengthy) key for each one. Allowing comma or colon-separated values would be great.

File: man/systemd-gpt-auto-generator.xml [hunk 8]
```diff
@@ -221,6 +221,11 @@
     systems, make sure to set the correct default subvolumes on them,
     using <command>btrfs subvolume set-default</command>.</para>
 
+    <para>If the system was booted via
+    <citerefentry><refentrytitle>systemd-stub</refentrytitle><manvolnum>7</manvolnum></citerefentry> and it
```
Comment (keszybz) [inline man/systemd-gpt-auto-generator.xml]: "it" is unclear here. Grammatically it *sounds* like a reference to "the system", but "the system" doesn't do this. It doesn't matter greatly, but I think it'd be nice to reword this to avoid this misleading phrasing.

"volume … will be … measured to PCR 15" — firstly, this is only if the volume is encrypted, right? Secondly, maybe say what is actually measured. Thirdly, maybe "into PCR 15" instead of "to", for consistency.

File: src/gpt-auto-generator/gpt-auto-generator.c [hunk 9]
```diff
@@ -89,7 +90,28 @@ static int add_cryptsetup(
...
+                /* We only measure the root volume key into PCR 15 if we are booted with sd-stub (i.e. in a
+                 * UKI), and sd-stub measured the UKI. We do this in order not to step into people's own PCR
+                 * assignment, under the assumption that people who are fine to use sd-stub with its PCR
+                 * assignments are also OK with our PCR 15 use here. */
+
+                r = efi_get_variable(EFI_LOADER_VARIABLE(StubPcrKernelImage), NULL, NULL, NULL); /* we don't actually care which PCR the UKI used for itself */
+                if (r == -ENOENT)
+                        log_debug_errno(r, "Will not measure volume key of volume '%s', because not booted via systemd-stub with measurements enabled.", id);
+                else if (r < 0)
+                        log_debug_errno(r, "Failed to determine whether booted via systemd-stub with measurements enabled, ignoring: %m");
```
Comment (keszybz) [inline src/gpt-auto-generator/gpt-auto-generator.c]: `log_warning_errno` ?

File: src/shared/generator.c [hunk 10]
```diff
@@ -110,6 +110,64 @@ int generator_add_symlink_full(
...
+        const char *to;
+        int r;
+
+        assert(dir);
+        assert(src);
+        assert(order);
+        assert(dst);
+
+        /* Adds in an explicit ordering dependency of type <order> from <src> to <dst>. If <instance> is
+         * specified, it is inserted into <fst> first as an instance */
```
Comment (keszybz) [inline src/shared/generator.c]: "`<dst>`", period.

File: src/basic/special.h [hunk 11]
```diff
@@ -87,6 +87,8 @@
 #define SPECIAL_REMOUNT_FS_SERVICE "systemd-remount-fs.service"
 #define SPECIAL_VOLATILE_ROOT_SERVICE "systemd-volatile-root.service"
 #define SPECIAL_UDEVD_SERVICE "systemd-udevd.service"
+#define SPECIAL_GROWFS_SERVICE "systemd-growfs@.service"
```
Comment (keszybz) [inline src/basic/special.h:90]: This is nice. I think we should rework the other places that generate a unit in full to have a common stub too. It's confusing to users if a unit doesn't have a static file anywhere.

File: units/systemd-pcrfs@.service.in [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+#  the Free Software Foundation; either version 2.1 of the License, or
+#  (at your option) any later version.
+
+[Unit]
+Description=TPM2 PCR File System Measurement of %f
+Documentation=man:systemd-pcrfs@.service(8)
+DefaultDependencies=no
+BindsTo=%i.mount
+Conflicts=shutdown.target
+After=%i.mount systemd-pcrfs-root.service
```
Comment (keszybz) [inline units/systemd-pcrfs@.service.in:16]: Hmm, so this would work best if we measure at most two file systems: root and something else. If we have more, than the measurement order will be racy, and we'll get different results…

File: docs/ENVIRONMENT.md [hunk 13]
```diff
@@ -506,7 +506,10 @@ SYSTEMD_HOME_DEBUG_SUFFIX=foo \
...
-`systemd-pcrphase`:
+`systemd-pcrphase`, `systemd-cryptsetup`:
 
-* `$SYSTEMD_PCRPHASE_STUB_VERIFY` – Takes a boolean. If false the requested
-  measurement is done even if no EFI stub usage was reported via EFI variables.
+* `$SYSTEMD_FORCE_MEASURE=1` — If set, force measuring of resources (which are
+  marked for measurement) even if not booted on a kernel equipped with
+  systemd-stub. Normally, requested measurement of resources is conditionalized
+  on kernels that have booted with `systemd-stub`. With this environment
+  variable the test for that my be bypassed, for testing purposes.
```
Comment (keszybz) [inline docs/ENVIRONMENT.md:515]: FWIW, this is an env var for which we don't provide stability guarantees, so we *could* change this even after a release.

File: src/cryptsetup/cryptsetup.c [hunk 14]
```diff
@@ -819,6 +820,14 @@ static int measure_volume_key(
                 return 0;
         }
 
+        r = efi_stub_measured();
+        if (r < 0)
+                return log_debug_errno(r, "Failed to detect if we are running on a kernel image with TPM measurement enabled: %m");
+        if (r == 0) {
+                log_debug("Kernel stub did not measure kernel image, skipping userspace measurement, too.");
```
Comment (keszybz) [inline src/cryptsetup/cryptsetup.c]: Hmm, strictly speaking "did not measure" is quite different from "did not measure into the expected PCR".

File: man/systemd-pcrphase.service.xml [hunk 15]
```diff
@@ -39,13 +45,23 @@
...
     <para>These services require
     <citerefentry><refentrytitle>systemd-stub</refentrytitle><manvolnum>7</manvolnum></citerefentry> to be
     used in a unified kernel image (UKI) setup. They execute no operation when invoked when the stub has not
     been used to invoke the kernel. The stub will measure the invoked kernel and associated vendor resources
-    into PCR 11 before handing control to it; once userspace is invoked these services then will extend
-    certain literal strings indicating various phases of the boot process into TPM2 PCR 11. During a regular
-    boot process the following strings are extended into PCR 11.</para>
+    into PCR 11 before handing control to it; once userspace is invoked the <literal>pcrphase</literal>
+    services then will extend certain literal strings indicating various phases of the boot process into TPM2
+    PCR 11. During a regular boot process the following strings are extended into PCR 11:</para>
```
Comment (keszybz) [inline man/systemd-pcrphase.service.xml]: "PCR 11 is extended with the following strings" ?

File: test/units/testsuite-70.sh [hunk 16]
```diff
@@ -172,6 +172,36 @@ else
     echo "/usr/lib/systemd/systemd-measure or PCR sysfs files not found, skipping signed PCR policy test case"
 fi
 
+if [ -e /usr/lib/systemd/systemd-pcrphase ] && \
+       [ -f /sys/class/tpm/tpm0/pcr-sha256/11 ]; then
+
+    # Let's measure the machine ID
+    tpm2_pcrread sha256:15 -Q -o /tmp/oldpcr15
+    mv /etc/machine-id /etc/machine-id.save
+    echo 994013bf23864ee7992eab39a96dd3bb > /etc/machine-id
```
Comment (keszybz) [inline test/units/testsuite-70.sh]: Please no space after redirection operator.

File: test/units/testsuite-70.sh [hunk 17]
```diff
@@ -172,6 +172,36 @@ else
...
+    SYSTEMD_FORCE_MEASURE=1 /usr/lib/systemd/systemd-pcrphase --machine-id
+    mv /etc/machine-id.save /etc/machine-id
+    tpm2_pcrread sha256:15 -Q -o /tmp/newpcr15
+
+    # And check it matches expectations
+    ( cat /tmp/oldpcr15 ;
+      ( echo -n "machine-id:994013bf23864ee7992eab39a96dd3bb" | openssl dgst -binary -sha256 ) ) | openssl dgst -binary -sha256 | cmp - /tmp/newpcr15
```
Comment (keszybz) [inline test/units/testsuite-70.sh]: That inner parentheses are not necessary, here and below.

**Review by keszybz — CHANGES_REQUESTED** [review]
Looks very nice. I think the user interface could be made nicer in one spot, which we should do before -rc1.

File: src/shared/tpm2-util.c (hunk 4 above)
Comment (poettering) [reply src/shared/tpm2-util.c]: subsystems might reuse the volume key for some other protection purposes (like we do here in a way), and they might need it shorter hence they might just hash it to make it so. If we send the same hash over the wire any such use of would leak to an attacker. So we accidentally broke the protection that others assumed the volume key has.

So, to protect against that we'll derive a hash from the volume key that is "private" to us, by HMACing some string that is specific to our use here. 

or in other words: sending the simple literal hash of the volume key over the wire should usually not be a problem – until some other subsystem uses the hash of the volume key for some entirely unrelated other purpose and then suddenly we are the avenue for an attack.

will extend the comment.

File: units/systemd-pcrfs@.service.in (hunk 12 above)
Comment (poettering) [reply units/systemd-pcrfs@.service.in:16]: that is absolutely true. i ignored the issue for now, as it doesn't matter for the gpt-auto case (where only the root fs and /var will be measured). it might matter for people who manually instantiate `systemd-pcrphase@.service` for some fs. but in that case they can quickly address this locally by creating a local drop-in that adds one After= and they are done.

if this becomes an issue we can certainly tweak this. For example I guess the `x-systemd.after=` mount option could be extended to also have an effect on the `systemd-pcrphase@.service` units or so. But I think this can be done later, this is technical enough to not require this right-away given that there's a relatively easy way to make it work via a drop-in that any person smart enough tto grok TPMs should be able to master, too.

File: man/crypttab.xml (hunk 7 above)
Comment (poettering) [reply man/crypttab.xml]: frankly, I don't expect anyone to use this. the automatic detection is what you want. this is in the territory of "let's make the common/default stuff easy, and the exotic stuff possible". This is exotic stuff...

comma separated is problematic because this is crypttab option sstring which is also comma separated.

anyway, I really wouldn't bother with this for now, it's a niche thing.

Comment (poettering) [reply man/crypttab.xml]: (but anyway, I added this now)

File: src/basic/special.h (hunk 11 above)
Comment (poettering) [reply src/basic/special.h:90]: which one do you have in mind?

**Comment by poettering** [comment]
made all suggested changes. ptal!

**Comment by poettering** [comment]
I'll merge this now. @keszybz review was primarily doc stuff, with some exceptions. The code changes in the most recent push were minimal, hence let's risk is. We can always do post merge fixups of course.

File: src/basic/special.h (hunk 11 above)
Comment (keszybz) [reply src/basic/special.h:90]: Basically anything that uses `generator_open_unit_file` would be a candidate. But I didn't look into it too closely. If the unit is not templated, we probably can't do that.
